port/osx/omrosdump.c

```diff
@@ -0,0 +1,482 @@
+			(thread_state_t)&threads[i].thread_state, &state_int_count);
+		if (KERN_SUCCESS != kr) {
+			goto done;
+		}
+		threads[i].cmdsize += state_int_count * sizeof(natural_t);
+		state_int_count = x86_FLOAT_STATE_COUNT;
+		kr = thread_get_state(thread_info[i], x86_FLOAT_STATE,
+			(thread_state_t)&threads[i].float_state, &state_int_count);
+		if (KERN_SUCCESS != kr) {
+			goto done;
+		}
+		threads[i].cmdsize += state_int_count * sizeof(natural_t);
+		state_int_count = x86_EXCEPTION_STATE_COUNT;
+		kr = thread_get_state(thread_info[i], x86_EXCEPTION_STATE,
+			(thread_state_t)&threads[i].exceptions, &state_int_count);
+		if (KERN_SUCCESS != kr) {
+			goto done;
+		}
+		threads[i].cmdsize += state_int_count * sizeof(natural_t);
+	}
+
+done:
+	if (KERN_SUCCESS == kr) {
+		*thread_commands = threads;
+	} else {
+		if (NULL != threads) {
+			free(threads);
+		}
+	}
+	for (i = 0; i < *thread_count; i++) {
+		mach_port_deallocate(mach_task_self(), thread_info[i]);
+	}
+	mach_vm_deallocate(mach_task_self(), (mach_vm_address_t)thread_info, (*thread_count) * sizeof(thread_act_t));
+	return kr;
+}
+
+static kern_return_t
+list_segment_commands(mach_port_t task_port, struct segment_command_64 **segment_commands, natural_t *segment_count)
+{
+	kern_return_t kr = KERN_SUCCESS;
+	struct segment_command_64 *segments = NULL;
+	struct segment_command_64 *tmp_segments = NULL;
+	vm_address_t address = 0;
+	uint32_t depth = 0;
+	vm_region_submap_info_data_64_t vm_info;
+	natural_t segment_array_size = 128;
+
+	*segment_count = 0;
+	segments = calloc(segment_array_size, sizeof(struct segment_command_64));
+	if (NULL == segments) {
+		kr = KERN_NO_SPACE;
+		goto done;
+	}
+
+	while (1) {
+		vm_size_t size = 0;
+		vm_address_t old_addr = address;
+		mach_msg_type_number_t info_count = VM_REGION_SUBMAP_INFO_COUNT_64;
+		vm_offset_t data_read = 0;
+		mach_msg_type_number_t data_size = 0;
+
+		if (*segment_count >= segment_array_size) {
+			segment_array_size += segment_array_size / 2;
+			tmp_segments = calloc(segment_array_size, sizeof(struct segment_command_64));
+			if (NULL == tmp_segments) {
+				kr = KERN_NO_SPACE;
+				goto done;
+			}
+			memcpy(tmp_segments, segments, *segment_count * sizeof(struct segment_command_64));
+			free(segments);
+			segments = tmp_segments;
+			tmp_segments = NULL;
+		}
+
+		kr = vm_region_recurse_64(task_port, &address, &size, &depth, (vm_region_info_t)&vm_info, &info_count);
+
+		if (KERN_SUCCESS != kr) { /* end of memory segments or actual error */
+			break;
+		}
+
+		if (vm_info.is_submap) {
+			depth += 1;
+			continue;
+		}
+
+		/* if we get the same region as previous, update previous region instead of creating a new one */
+		if ((*segment_count > 0) && (old_addr == (address + segments[*segment_count - 1].vmsize))) {
+			*segment_count -= 1;
+			if (size <= segments[*segment_count].vmsize) {
+				size = segments[*segment_count].vmsize;
+				address += 1; /* so we don't endlessly loop on one region */
```

Comment:
This appears to have the potential to have an `address` value that doesn't refer to any segment.
Perhaps this increment should only occur if `size` is zero? If `size` is non-zero, `address` will be updated below (on line 367).